### PR TITLE
Literal generation for UDTs

### LINKED ISSUE #12192: Scaffolding literals for SqlHierarchyId (and possibly other UDTs) doesn't work
Originally posted as https://github.com/aspnet/EntityFrameworkCore/issues/10108#issuecomment-393276024 by @KyleTraynor

When I try to use SqlHierarchyId in one of my classes and then add-migration, i get the error:
The current CSharpHelper cannot scaffold literals of type 'Microsoft.SqlServer.Types.SqlHierarchyId'. Configure your services to use one that can.

I am using EF Core 2.1 with ASPNetCore 2.1 while targeting .net 4.7.1

Any ideas what I need to do to add a column that uses SqlHierarchyId to my existing entity?

Thanks

### PR DESCRIPTION
**Pull request by ajcvickers**
Fixes #12192

### CODE REVIEW

File: test/EFCore.Design.Tests/Design/Internal/CSharpHelperTest.cs [hunk 1]
```diff
@@ -522,6 +521,55 @@ public void Literal_with_unsupported_node_throws()
...
+        [Fact]
+        public virtual void Can_generate_SqlHierarchyId_literal()
+        {
+            var typeMapping = CreateTypeMappingSource(
+                new TestTypeMappingPlugin<FakeSqlHierarchyId>(
+                    SqlServerUdtTypeMapping.CreateSqlHierarchyIdMapping(
+                        typeof(FakeSqlHierarchyId)).LiteralGenerator));
+
+            Assert.Equal(
+                "Microsoft.SqlServer.Types.FakeSqlHierarchyId.Parse(new System.Data.SqlTypes.SqlString(\"/1/1/3/\"))",
```
Comment (bricelam) [inline test/EFCore.Design.Tests/Design/Internal/CSharpHelperTest.cs]: I think there's an implicit conversion available from `string` to `SqlString`...

File: test/EFCore.Design.Tests/Design/Internal/CSharpHelperTest.cs [hunk 2]
```diff
@@ -522,6 +521,55 @@ public void Literal_with_unsupported_node_throws()
...
+        {
+            var typeMapping = CreateTypeMappingSource(
+                new TestTypeMappingPlugin<FakeSqlGeometry>(
+                    SqlServerUdtTypeMapping.CreateSqlSpatialMapping(
+                            typeof(FakeSqlGeometry),
+                            "geometry")
+                        .LiteralGenerator));
+
+            Assert.Equal(
+                "Microsoft.SqlServer.Types.FakeSqlGeometry.Deserialize(new System.Data.SqlTypes.SqlBytes(new byte[] { 0, 0, 0, 0, 1, 12, 0, 0, 0, 0, 0, 0, 240, 63, 0, 0, 0, 0, 0, 0, 0, 64 }))",
```
Comment (bricelam) [inline test/EFCore.Design.Tests/Design/Internal/CSharpHelperTest.cs]: Oh I'd much prefer `SqlGeometry.STGeomFromText("wkt", srid)`

Comment (bricelam) [reply test/EFCore.Design.Tests/Design/Internal/CSharpHelperTest.cs]: (And *I* actually know what those bytes mean 😉)

Comment (bricelam) [reply test/EFCore.Design.Tests/Design/Internal/CSharpHelperTest.cs]: NB: Use `.AsTextZM()` to get the right WKT value.

File: test/EFCore.Design.Tests/Design/Internal/CSharpHelperTest.cs (hunk 1 above)
Comment (ajcvickers) [reply test/EFCore.Design.Tests/Design/Internal/CSharpHelperTest.cs]: @bricelam There is, but then I would have to figure out how to represent and translate that in the expression tree, and so for now I went with something super simple. Same for implicit conversion from `SqlBytes` to `byte[]`.

File: test/EFCore.Design.Tests/Design/Internal/CSharpHelperTest.cs (hunk 2 above)
Comment (ajcvickers) [reply test/EFCore.Design.Tests/Design/Internal/CSharpHelperTest.cs]: I wasn't sure if anything other than serialize/deserialize would be lossy, but I can change to this if it is safe.

Comment (bricelam) [reply test/EFCore.Design.Tests/Design/Internal/CSharpHelperTest.cs]: Yep, this is non-lossy and human readable.

**Pull request merged**